File: jserver-security/src/main/java/com/github/gr3gdev/jserver/security/TokenServerPlugin.java
package com.github.gr3gdev.jserver.security;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.gr3gdev.jserver.plugin.ServerPlugin;
import com.github.gr3gdev.jserver.security.user.UserData;
import io.fusionauth.jwt.Signer;
import io.fusionauth.jwt.domain.JWT;
import io.fusionauth.jwt.rsa.RSASigner;

import java.security.PrivateKey;
import java.time.ZonedDateTime;
import java.time.temporal.ChronoUnit;
import java.util.UUID;

/**
 * TokenServerPlugin.
 *
 * @author Gregory Tardivel
 */
public class TokenServerPlugin implements ServerPlugin {

    private PrivateKey key;
    private String id = UUID.randomUUID().toString();
    private String issuer = "Unknown";

    public TokenServerPlugin(final PrivateKey privateKey) {
        this.key = privateKey;
    }

    /**
     * Specify JWT id.
     *
     * @param id JWT id
     * @return TokenServerPlugin
     */
    public TokenServerPlugin id(final String id) {
        this.id = id;
        return this;
    }

    /**
     * Specify JWT issuer.
     *
     * @param issuer JWT issuer
     * @return TokenServerPlugin
     */
    public TokenServerPlugin issuer(final String issuer) {
        this.issuer = issuer;
        return this;
    }

    /**
     * Create a JWT from user data.
     */
    public <T extends UserData> String createToken(T userData, Long expirationMillis) {
        final Signer signer = RSASigner.newSHA256Signer(this.key);
        final ZonedDateTime now = ZonedDateTime.now();

        final ObjectMapper mapper = new ObjectMapper();
        // Set the JWT Claims
        try {
            final JWT jwt = new JWT()
                    .setUniqueId(this.id)
                    .setIssuer(this.issuer)
                    .setIssuedAt(now)
                    .setSubject(mapper.writeValueAsString(userData));

            // Add the expiration
            if (expirationMillis > 0) {
                jwt.setExpiration(now.plus(expirationMillis, ChronoUnit.MILLIS));
            }
            return JWT.getEncoder().encode(jwt, signer);
        } catch (JsonProcessingException e) {
            e.printStackTrace();
        }
        return "";
    }

}

File: jserver-security/src/main/kotlin/com/github/gr3gdev/jserver/security/TokenServerPlugin.kt
package com.github.gr3gdev.jserver.security

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.github.gr3gdev.jserver.plugin.ServerPlugin
import com.github.gr3gdev.jserver.security.user.UserData
import io.jsonwebtoken.JwtBuilder
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import java.io.InputStream
import java.security.Key
import java.security.KeyStore
import java.util.*

class TokenServerPlugin(alias: String, keystoreFile: InputStream, keystorePassword: CharArray) : ServerPlugin {

    private val mapper = jacksonObjectMapper()
    private val key: Key
    private var id: String = UUID.randomUUID().toString()
    private var issuer = "Unknown"

    init {
        keystoreFile.use {
            val keystore = KeyStore.getInstance(KeyStore.getDefaultType())
            keystore.load(it, keystorePassword)
            this.key = keystore.getKey(alias, keystorePassword)
        }
    }

    /**
     * Specify JWT id.
     */
    fun id(id: String): TokenServerPlugin {
        this.id = id
        return this
    }

    /**
     * Specify JWT issuer.
     */
    fun issuer(issuer: String): TokenServerPlugin {
        this.issuer = issuer
        return this
    }

    /**
     * Create a JWT from user data.
     */
    fun <T : UserData> createToken(userData: T, expirationMillis: Long): String {
        val nowMillis = System.currentTimeMillis()
        val now = Date(nowMillis)

        // Set the JWT Claims
        val builder: JwtBuilder = Jwts.builder()
                .setId(id)
                .setIssuedAt(now)
                .setSubject(mapper.writeValueAsString(userData))
                .setIssuer(issuer)
                .signWith(SignatureAlgorithm.RS512, key)

        // Add the expiration
        if (expirationMillis > 0) {
            val expMillis: Long = nowMillis + expirationMillis
            val exp = Date(expMillis)
            builder.setExpiration(exp)
        }
        return builder.compact()
    }

}